src/crimson/os/seastore/randomblock_manager.h

```diff
@@ -0,0 +1,365 @@
+
+  DENC(rbm_bitmap_block_t, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.header, p);
+    denc(v.buf, p);
+    DENC_FINISH(p);
+  }
+};
+
+}
+
+WRITE_CLASS_DENC_BOUNDED(
+  crimson::os::seastore::rbm_metadata_header_t
+)
+WRITE_CLASS_DENC_BOUNDED(
+  crimson::os::seastore::rbm_bitmap_block_t
+)
+WRITE_CLASS_DENC_BOUNDED(
+  crimson::os::seastore::rbm_bitmap_block_header_t
+)
+
+namespace crimson::os::seastore {
+
+class RandomBlockManager {
+public:
+
+  struct mkfs_config_t {
+    std::string path;
+    blk_paddr_t start;
+    blk_paddr_t end;
+    size_t block_size = 0;
+    size_t total_size = 0;
+    seastore_meta_t meta;
+  };
+  using mkfs_ertr = crimson::errorator<
+	crimson::ct_error::input_output_error,
+	crimson::ct_error::invarg           
+	>;
+  /*
+   * Ondisk layout
+   *
+   * -------------------------------------------------------------------------------------------------
+   * | rbm_metadata_header_t | rbm_bitmap_block_t 1 | rbm_bitmap_block_t 2 | ... |    data blocks    |
+   * -------------------------------------------------------------------------------------------------
+   */
+
+  mkfs_ertr::future<> mkfs(mkfs_config_t);
+
+  using read_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::invarg,
+    crimson::ct_error::enoent,
+    crimson::ct_error::erange>;
+  read_ertr::future<> read(uint64_t addr, bufferptr &buffer);
+
+  using write_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error, 
+    crimson::ct_error::invarg,             
+    crimson::ct_error::ebadf,              
+    crimson::ct_error::enospc, 
+    crimson::ct_error::erange
+    >;
+  write_ertr::future<> write(uint64_t addr, bufferptr &buf);
+
+  using open_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::invarg,
+    crimson::ct_error::enoent>;
+  open_ertr::future<> open(const std::string &path, blk_paddr_t start);
+
+  using close_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::invarg>;
+  close_ertr::future<> close();
+
+  using allocate_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error, 
+    crimson::ct_error::invarg,             
+    crimson::ct_error::enospc              
+    >;
+  using allocate_ret = allocate_ertr::future<
+    std::optional<bufferptr>
+    >;
+  /*
+   * alloc_extent
+   *
+   * The role of this function is to find out free blocks the transaction requires.
+   * To do so, alloc_extent() looks into both in-memory allocator and freebitmap blocks.
+   * But, in-memory allocator is the future work, and is not implemented yet,
+   * we use freebitmap directly to allocate freeblocks for now.
+   * 
+   * Each bit in freebitmap block represents whether a block is allocated or not.
+   *
+   * TODO: multiple allocation
+   *
+   */
+  allocate_ertr::future<> alloc_extent(Transaction &t, size_t size); // allocator, return blocks
```

Comment:
Could you give an example? alloc_extent() in randomblockmanager probably be called by cache->alloc_extent() to find the empty space and physical address.

Reply:
Agree with you.

In the current seastore, the actual physical location (paddr) of the new extents are only available (or allocated) after the transaction/record is written by `Journal::submit_record()`, because they are always imbedded in the transactional record. See https://github.com/ceph/ceph/blob/33e02a9da342b6bea28f9c44c465ed847840c7e4/src/crimson/os/seastore/transaction_manager.cc#L245-L250

So I don't think there is any suitable example for the extents that need to be allocated prior to transaction submission. And I think this is also why seastore components need to be generalized in order to support random block devices.

Reply:
@xxhdx1985126  is working on the TransactionManager changes for that now.

Reply:
IIUC, the current segment cleaner would make sure that there's always enough empty space on disk for journal. So, following the philosophy of the current seastore design, I think maybe, in the future, we should implement some kind of device space cleaner that takes the responsibility for making sure there's enough space on any devices that would be holding the data of the submitting transaction(probably by the help of the manager of those devices), and it will be the device managers' responsibility to find the correct empty space for the journal/block write. I think we can postpone the actual allocation of extents to the point of time when actual block writes happen.

Note that, according to discussions [here](https://docs.google.com/document/d/1brbZ73sPBKbbsLteiuaNGRZ88G8Vp91MdnyORcb7KyM/edit#), we need to preserve the `alloc_extent` method in upper layers for the sake of PMEM support.

So, I think maybe we can merge this `RandomBlockManager::alloc_extent` with `RandomBlockManager::write`, and make sure there's enough facilities in `RandomBlockManager` for the future block device space cleaner to clean the corrresponding random block device.

Am I right? :-)

Reply:
>  I think we can postpone the actual allocation of extents to the point of time when actual block writes happen.

I think not, this is the special case for the segmented backend because LBA tree will store relative physical addresses of the record-imbedded fresh extents before the transaction is submitted and written. For random blocks there are no relative physical addresses because they are not imbedded in the record, their physical locations need to be allocated from the in-memory allocator before updating the LBA tree, which are supposed to be done in `Cache::alloc_new_extent()` and `RandomBlockManager::alloc_extent`. See https://github.com/ceph/ceph/blob/4c0f3127b195df3538109ca9ca869a07219013e6/src/crimson/os/seastore/transaction_manager.h#L257-L265

Reply:
Oh, yes, that's right.

Reply:
Oh, yes, that's right.

Reply:
It's not a special case for the segmented backend -- it's specific to extents allocated as part of the journal writes.  There are reasons to do that even with a circular journal and RBM.  extents expected to be short lived can be combined with the journal write to avoid an extra IO.  We will want to preserve this behavior regardless of the backend.

For extents not expected to be short lived, we'll instead choose to write directly to an RBM location OR a segment other than the journal segment.  @xxhdx1985126 's TransactionManager changes will add interface support for that kind of write, which will be backed either by a seperate segment or by this new RBM allocate mechanism.

Reply:
That is, there are two kinds of extent:
1) Extent written as part of the journal record (record_t::extents) where we don't know the prior to commit time
2) Extent written directly to final location (new TransactionManager interface).  Write happens prior to final commit, address is known prior to commit time.

Type 2 extents may be backed by pmem, a segment other than the journal segment, or by an extent in the RBM.

Does that make sense?

Reply:
> That is, there are two kinds of extent:
> 
> 1. Extent written as part of the journal record (record_t::extents) where we don't know the prior to commit time
> 2. Extent written directly to final location (new TransactionManager interface).  Write happens prior to final commit, address is known prior to commit time.
> 
> Type 2 extents may be backed by pmem, a segment other than the journal segment, or by an extent in the RBM.
> 
> Does that make sense?

I think so:-)

Reply:
I agree with the 2 kinds of extents.

> There are reasons to do that even with a circular journal and RBM. Extents expected to be short lived can be combined with the journal write to avoid an extra IO. 

Not really sure about this until we can actually test with the random-write optimized block devices. But I believe we could make it a tunable about whether to imbed the short lived extents in the circular journal with RBM.

Reply:
Yep, that's the idea.